amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+
+        if names:
+            for key in names:
+                assert isinstance(key, float)
+                assert isinstance(names[key], str)
+
+
+# ------------------------------------------------------------------------------
+
+
+def starts_from_time_signature(
+    ts_str: str, enforce_2s_3s: bool = True, minimum_pulse: int = 64
+) -> list:
+    """
+    Create a start hierarchy for almost any time signature
+    directly from a string (e.g., "4/4") without dependencies.
+
+
+    Parameters
+    ----------
+    ts_str
+        Any valid string repersenting a time signature. See examples below.
+    enforce_2s_3s
+        Map 4 to 2+2 and 6 to 3+3 etc. following the conventions of those time signatures.
+    minimum_pulse
+        Recursively create further down to the denominator level specified. Defaults to 64 for 64th notes.
+
+    Returns
+    -------
+    list
+        Returns a list of lists with start positions by level.
+
+    Examples
+    --------
+    The following examples index the most interesting level:
+
+    >>> starts_from_time_signature("4/4")[1]  # note the half cycle division
+    [0.0, 2.0, 4.0]
+
+    >>> starts_from_time_signature("4/4", enforce_2s_3s=False)[1]  # note the absence of a half cycle division
+    [0.0, 1.0, 2.0, 3.0, 4.0]
+
+    >>> starts_from_time_signature("2/2")[1]  # note the presence of a half cycle division
+    [0.0, 2.0, 4.0]
+
+    >>> starts_from_time_signature("6/8")[1]  # note the macro-beat division
+    [0.0, 1.5, 3.0]
+
+
+    Numerators like 5 and 7 are supported.
+    Use the total value only to avoid segmentation about the denominator level:
+
+    >>> starts_from_time_signature("5/4")[1]  # note no 2+3 or 3+2 level division
+    [0.0, 1.0, 2.0, 3.0, 4.0, 5.0]
+
+    Or use numerator addition in the form "X+Y/Z" to clarify this level ...
+
+    >>> starts_from_time_signature("2+3/4")[1]  # note the 2+3 division
+    [0.0, 2.0, 5.0]
+
+    >>> starts_from_time_signature("2+2+3/8")[1]  # note the 2+2+3 division
+    [0.0, 1.0, 2.0, 3.5]
+
+    >>> starts_from_time_signature("2+2+2+3/8")[1]  # note the 2+2+2+3 division
+    [0.0, 1.0, 2.0, 3.0, 4.5]
+
+    Only standard denominators are supported:
+    i.e., time signatures in the form X/ one of
+    1, 2, 4, 8, 16, 32, or 64.
+    No so-called "irrational" meters yet (e.g., 2/3), sorry!
+
+    Likewise, the minimum_pulse must be set to one of these values
+    (default = 64 for 64th note) and not be longer than the meter.
+
+    Finally, although we support and provide defaults for time signatures in the form "2+3/8",
+    there is no such support for more than one "/" (i.e., the user must build cases like "4/4 + 3/8" explicitly).
+    """
+
+    # Prep and checks
+    numerator, denominator = ts_str.split("/")
+    numerators = [
+        int(x) for x in numerator.split("+")
+    ]  # Note: support "+" before one "/", e.g., `2+3/4`
+    denominator = int(denominator)
+
+    supported_denominators = [1, 2, 4, 8, 16, 32, 64]
```

Comment:
Is there a rationale for encoding 64 as the maximum value here? Can't we programmatically respond to greater values of minimum_pulse?

Reply:
Programmatically, we could indeed respond to greater values, but there is psychological evidence for structures beyond this ceasing to be 'meter' in the same sense. While there was once a practice of extending 'metrical structure' into the realm of 'form' (e.g., Cooper and Meyer), that seems to be a poor reflection of the psychology of meter (see e.g., London _Zeitnetz_).

I consider that reason enough and note that this is more principled than other restrictions: we're also limited to single metres (not mixtures that would make a case for 96 divisions of the bar, for instance), and to exclude the inaptly named 'irrational' metres (e.g., `X/6`). 

I've added a comment on the docs, but a shorter one than this ;)